relentless/simulate/lammps.py

```diff
@@ -0,0 +1,870 @@
+                                                                    ftol=self.force_tolerance,
+                                                                    maxiter=self.max_iterations,
+                                                                    maxeval=100*self.max_iterations)]
+
+        return cmds
+
+class AddLangevinIntegrator(LAMMPSOperation):
+    """Langevin dynamics for a NVE ensemble.
+
+    Parameters
+    ----------
+    dt : float
+        Time step size for each simulation iteration.
+    friction : float or dict
+        Drag coefficient for each particle type (shared or per-type).
+    seed : int
+        Seed used to randomly generate a uniform force.
+
+    """
+    def __init__(self, dt, friction, seed):
+        self.friction = friction
+        self.seed = seed
+
+        self._fix_langevin = self.new_fix_id()
+        self._fix_nve = self.new_fix_id()
+
+    def to_commands(self, sim):
+        """Calls the appropriate LAMMPS commands for the Langevin integrator operation.
+
+        Parameters
+        ----------
+        sim : :py:class:`Simulation`
+            The simulation object.
+
+        Returns
+        -------
+        array_like
+            The LAMMPS commands for this operation.
+
+        Raises
+        ------
+        ValueError
+            If the simulation ensemble does not have constant V.
+        ValueError
+            If the per-type particle masses are not set.
+        KeyError
+            If the friction factor for all types are not set (if per-type
+            friction factors are given).
+
+        """
+        if not sim.ensemble.constant['V']:
+            raise ValueError('Simulation ensemble does not have constant V.')
+
+        # obtain per-type mass (arrays 1-indexed using lammps convention)
+        Ntypes = len(sim.ensemble.types)
+        mass = np.squeeze(sim.lammps.numpy.extract_atom('mass'))
+        if mass is None or mass.shape != (Ntypes+1,):
+            raise ValueError('Per-type masses not set.')
+
+        # obtain per-type friction factor
+        friction = np.zeros_like(mass)
+        for t in sim.ensemble.types:
+            try:
+                friction[sim.type_map[t]] = self.friction[t]
+            except TypeError:
+                friction[sim.type_map[t]] = self.friction
+            except KeyError:
+                raise KeyError('The friction factor for type {} is not set.'.format(t))
+
+        # compute per-type damping parameter and rescale if multiple types
+        damp = np.divide(mass, friction, where=(friction>0))
+        damp_ref = damp[1]
+        if Ntypes>1:
+            scale = damp/damp_ref
+            scale_str = ' '.join(['scale {} {}'.format(i+1,s) for i,s in enumerate(scale[2:])])
+        else:
+            scale_str = ''
+
+        cmds = ['fix {idx} {group_idx} langevin {t_start} {t_stop} {damp} {seed} {scaling}'.format(idx=self._fix_langevin,
```

Comment:
This is a pedantic point, and I'm not sure if it even matters, but can you put the NVE fix before the langevin fix to be safe?